plugins/janus_sip.c

```diff
@@ -4467,39 +4471,68 @@ static void *janus_sip_handler(void *data) {
-			if(session->callee == NULL) {
-				janus_mutex_unlock(&session->mutex);
-				JANUS_LOG(LOG_ERR, "Wrong state (no callee?)\n");
-				error_code = JANUS_SIP_ERROR_WRONG_STATE;
-				g_snprintf(error_cause, 512, "Wrong state (no callee?)");
-				goto error;
-			}
-			janus_mutex_unlock(&session->mutex);
 			JANUS_VALIDATE_JSON_OBJECT(root, sipmessage_parameters,
-				error_code, error_cause, TRUE,
-				JANUS_SIP_ERROR_MISSING_ELEMENT, JANUS_SIP_ERROR_INVALID_ELEMENT);
+			      	error_code, error_cause, TRUE,
+			      	JANUS_SIP_ERROR_MISSING_ELEMENT, JANUS_SIP_ERROR_INVALID_ELEMENT);
 			if(error_code != 0) {
-				janus_mutex_unlock(&session->mutex);
 				goto error;
 			}
+			gboolean in_dialog_message = TRUE;
+			json_t *uri = json_object_get(root, "uri");
+			const char *uri_text = json_string_value(uri);
+			if(uri != NULL)
+				in_dialog_message = FALSE;
+
+			if(in_dialog_message) {
+				if(!(session->status == janus_sip_call_status_inviting || janus_sip_call_is_established(session))) {
+					JANUS_LOG(LOG_ERR, "Wrong state (not established? status=%s)\n", janus_sip_call_status_string(session->status));
+					g_snprintf(error_cause, 512, "Wrong state (not in a call?)");
+					goto error;
+				}
+				janus_mutex_lock(&session->mutex);
+				if(session->callee == NULL) {
+					janus_mutex_unlock(&session->mutex);
+					JANUS_LOG(LOG_ERR, "Wrong state (no callee?)\n");
+					error_code = JANUS_SIP_ERROR_WRONG_STATE;
+					g_snprintf(error_cause, 512, "Wrong state (no callee?)");
+					goto error;
+				}
+				janus_mutex_unlock(&session->mutex);
+			} else {
+				if(session->account.registration_status != janus_sip_registration_status_registered &&
+				   session->account.registration_status != janus_sip_registration_status_disabled) {
+					JANUS_LOG(LOG_ERR, "Wrong state (not registered)\n");
+					error_code = JANUS_SIP_ERROR_WRONG_STATE;
+					g_snprintf(error_cause, 512, "Wrong state (not registered)");
+					goto error;
+				}
+				janus_sip_uri_t target_uri;
+				if(janus_sip_parse_uri(&target_uri, uri_text) < 0) {
+					JANUS_LOG(LOG_ERR, "Invalid user address %s\n", uri_text);
+					error_code = JANUS_SIP_ERROR_INVALID_ADDRESS;
+					g_snprintf(error_cause, 512, "Invalid user address %s\n", uri_text);
+					goto error;
+				}
+			}
 
 			const char *content_type = "text/plain";
 			json_t *content_type_text = json_object_get(root, "content_type");
 			if(content_type_text && json_is_string(content_type_text))
 				content_type = json_string_value(content_type_text);
 
 			const char *msg_content = json_string_value(json_object_get(root, "content"));
-			nua_message(session->stack->s_nh_i,
-				SIPTAG_CONTENT_TYPE_STR(content_type),
-				SIPTAG_PAYLOAD_STR(msg_content),
-				TAG_END());
+			if(in_dialog_message) {
+				nua_message(session->stack->s_nh_i,
+					SIPTAG_CONTENT_TYPE_STR(content_type),
+					SIPTAG_PAYLOAD_STR(msg_content),
+					TAG_END());
+			} else {
+				nua_message(session->stack->s_nh_r,
```

Comment:
I don't think this will work. First of all, this assumes a dialog for the REGISTER exists for that session, which will not be the case if someone registered as a guest (and maybe if it's a helper too? but I'd have to check). Besides, not sure if it's safe to re-use that dialog anyway. Maybe a new dialog should be created for the job, and torn down when it succeeds?

Reply:
_"First of all, this assumes a dialog for the REGISTER exists for that session, which will not be the case if someone registered as a guest (and maybe if it's a helper too? but I'd have to check). "_
You are right. I will update PR.

 _"Maybe a new dialog should be created for the job, and torn down when it succeeds?"_
Do you mean a new nua_handle since SIP dialog is not created for SIP MESSAGE?
Not sure if we really need to destroy nua_handle when the message is sent or we can use the same nua_handle for all SIP MESSAGEs?

Reply:
> Not sure if we really need to destroy nua_handle when the message is sent or we can use the same nua_handle for all SIP MESSAGEs?

That might be an option too, and easier to clean up too :+1:

Reply:
Thanks. I updated the pull request.